File: javascript/387.first-unique-character-in-a-string.js
/*
 * @lc app=leetcode id=387 lang=javascript
 *
 * [387] First Unique Character in a String
 */

// @ts-check

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var firstUniqChar = function (s) {
  /** @type {Set<string>} */
  const set = new Set();
  /** @type {Map<string, number>} */
  const map = new Map();
  for (let i = 0; i < s.length; i++) {
    if (set.has(s[i])) {
      map.delete(s[i]);
    } else {
      set.add(s[i]);
      map.set(s[i], i);
    }
  }
  return map.size > 0 ? map.values().next().value : -1;
};
// @lc code=end

File: rust/src/bin/387.first-unique-character-in-a-string.rs
/*
 * @lc app=leetcode id=387 lang=rust
 *
 * [387] First Unique Character in a String
 */

pub struct Solution;

// @lc code=start
impl Solution {
    pub fn first_uniq_char(s: String) -> i32 {
        use std::collections::{BTreeMap, HashSet};

        let len = s.len();
        let bytes = s.as_bytes();
        let mut set = HashSet::new();
        let mut map = BTreeMap::new();

        for i in 0..len {
            if set.contains(&bytes[i]) {
                map.remove(&bytes[i]);
            } else {
                set.insert(bytes[i]);
                map.insert(bytes[i], i);
            }
        }

        if let Some((_, &i)) = map.iter().next() {
            i as i32
        } else {
            -1
        }
    }
}
// @lc code=end
